# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1259949341 18000
#      Fri Dec 04 12:55:41 2009 -0500
# Node ID 5cd48568ec591027e84571371b5734805cd9f26c
# Parent  e9fe12c1b0c0bf01685647d67325404c8c08697f
src/Makefile.am: generate output in mkdefs script, not Makefile

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,14 @@
+2009-12-04  John W. Eaton  <jwe@octave.org>
+
+	* Makefile.am (mkdefs): New arguments, srcdir and file.
+	Generate output here.
+	(.cc.df): Not here.
+
 2009-12-04  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.am (BUILT_SOURCES): Include graphics-props.cc in the list.
 
 2009-12-04  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.am (liboctinterp_la_SOURCES): Remove $(OPT_HANDLERS)
 	from the list.
diff --git a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -87,26 +87,19 @@ EXTRA_DIST = \
   octave.gperf \
   version.h.in \
   $(BUILT_DISTFILES)
 
 DLL_CDEFS = @OCTINTERP_DLL_DEFS@
 DLL_CXXDEFS = @OCTINTERP_DLL_DEFS@
 
 .cc.df:	
-	(echo "// DO NOT EDIT!  Generated automatically by mkdefs." ; \
-	  case "$<" in \
-	    $(srcdir)/*) xsrcdir="`echo $< | sed 's,^$(srcdir),src,'`" ;; \
-	    *) xsrcdir="src/$<" ;; \
-	  esac ; \
-	  echo " XDEFUN_FILE_NAME (\"$$xsrcdir\")" ; \
-	  egrep '^(///*|/\*) *PKG_ADD:' $< ; \
-	  $(CXXCPP) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) \
-	    $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) \
-	    -DMAKE_BUILTINS $< | $(srcdir)/mkdefs) > $@-t
+	$(CXXCPP) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) \
+	  $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) \
+	  -DMAKE_BUILTINS $< | $(srcdir)/mkdefs $(srcdir) $< > $@-t
 	mv $@-t $@
 
 OPT_HANDLERS = \
   DASPK-opts.cc \
   DASRT-opts.cc \
   DASSL-opts.cc \
   LSODE-opts.cc \
   Quad-opts.cc
diff --git a/src/mkdefs b/src/mkdefs
--- a/src/mkdefs
+++ b/src/mkdefs
@@ -1,11 +1,11 @@
 #! /bin/sh
 #
-# Copyright (C) 1996, 1999, 2000, 2003, 2006, 2007 John W. Eaton
+# Copyright (C) 1996, 1999, 2000, 2003, 2006, 2007, 2009 John W. Eaton
 #
 # This file is part of Octave.
 # 
 # Octave is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License as published by the
 # Free Software Foundation; either version 3 of the License, or (at
 # your option) any later version.
 # 
@@ -15,16 +15,36 @@
 # for more details.
 # 
 # You should have received a copy of the GNU General Public License
 # along with Octave; see the file COPYING.  If not, see
 # <http://www.gnu.org/licenses/>.
 
 SED=${SED:-'sed'}
 
+if [ $# -ne 2 ]; then
+  echo "usage: mkdefs srcdir file < preprocessed-file-contents" 1>&2
+  exit 1
+fi
+
+srcdir="$1"
+file="$2";
+
+echo "// DO NOT EDIT!  Generated automatically by mkdefs."
+case "$file" in
+  $srcdir/*)
+    xsrcdir=`echo "$file" | sed "s,^$srcdir,src,"`
+  ;;
+  *)
+    xsrcdir="src/$file"
+  ;;
+esac
+echo " XDEFUN_FILE_NAME (\"$xsrcdir\")"
+egrep '^(///*|/\*) *PKG_ADD:' $file
+
 $SED -n -e '/^ *BEGIN_INSTALL_BUILTIN/{
   : loop
   s/BEGIN_INSTALL_BUILTIN//
   s/^#[ \t][ \t]*[0-9][0-9]*.*$//
   /END_INSTALL_BUILTIN/b done
   p
   n
   b loop
